xarray/core/common.py

```diff
@@ -135,24 +136,49 @@ def wrapped_func(self, dim=None, **kwargs):  # type: ignore[misc]
     ).strip()
 
 
-class AbstractArray:
+class AbstractArray(metaclass=ABCMeta):
     """Shared base class for DataArray and Variable."""
 
     __slots__ = ()
 
-    def __bool__(self: Any) -> bool:
+    @property
+    @abstractmethod
+    def dims(self) -> tuple[Hashable, ...]:
+        ...
+
+    @property
+    @abstractmethod
+    def shape(self) -> tuple[int, ...]:
+        ...
+
+    @property
+    @abstractmethod
+    def ndim(self) -> int:
+        ...
+
+    @property
+    @abstractmethod
+    def values(self) -> np.ndarray:
+        ...
+
+    @property
+    @abstractmethod
+    def data(self) -> Any:  # TODO: type me?
```

Comment:
The informal guarantee is that `.data` is [numpy duck array](https://numpy.org/neps/nep-0022-ndarray-duck-typing-overview.html), i.e., something with an interface matching `np.ndarray`. Unfortunately, this is a somewhat ill-defined notation at present from a type-system perspective.

In practice, Xarray only guarantees that `.data` passes these checks:
https://github.com/pydata/xarray/blob/63ba862d03c8d0cd8b44d2071bc360e9fed4519d/xarray/core/utils.py#L259-L270

Reply:
I think again we could introduce some `Protocol` here, a bit messy with the two special cases of array_namespace vs function but maybe simply a Union of two Protocols could work.

Reply:
Ok, maybe such a simple Protocol will not solve the typing issues here, since we probably expect that it supports most numpy ndarray methods...

Reply:
Coming back to this problem.
I think the really best solution would be to make `DataArryay` a Generic Type which indicates the underlying datastructure (something like `DataArray[np.ndarray]`).

Only two problems:

there is no way to represent the data type of coordinates independently and also for Datasets things don't work if the underlying data type is mixed.

This kind of typing would make it more convoluted for beginners (maybe it is ok though).

Maybe I open an extra issue for this.